File: samples/AccountSample/src/main/java/com/daya/android/account/sample/MainActivity.java
package com.daya.android.account.sample;

import android.Manifest;
import android.accounts.Account;
import android.accounts.AccountManager;
import android.accounts.AccountManagerFuture;
import android.accounts.AuthenticatorException;
import android.accounts.OperationCanceledException;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import java.io.IOException;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private static final int PERMISSION_REQUEST_CODE = 1000;
    private static final int CHOOSE_ACCOUNT_REQUEST_CODE = 2000;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        String permission = Manifest.permission.GET_ACCOUNTS;
        if (ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED) {
            Log.d(TAG, permission + " permission granted.");

            chooseAccountIntent();
        } else {
            Log.w(TAG, permission + " permission denied.");

            if (ActivityCompat.shouldShowRequestPermissionRationale(this, permission)) {
                Log.d(TAG, "shouldShowRequestPermissionRationale: true");
                ActivityCompat.requestPermissions(this, new String[]{permission}, PERMISSION_REQUEST_CODE);
            } else {
                Log.d(TAG, "shouldShowRequestPermissionRationale: false");
                ActivityCompat.requestPermissions(this, new String[]{permission}, PERMISSION_REQUEST_CODE);
            }
        }


    }

    private void chooseAccountIntent() {
        Intent intent = AccountManager.newChooseAccountIntent(
                null, null, new String[]{"com.google"}, null, null, null, null);
        startActivityForResult(intent, CHOOSE_ACCOUNT_REQUEST_CODE);
    }

    private void readAccount() {
        AccountManager accountManager = AccountManager.get(this);
        Account[] accounts = accountManager.getAccounts();

        for (Account account : accounts) {
            String accountName = account.name;
            String accountType = account.type;

            Log.d(TAG, "Account Name: " + accountName);
            Log.d(TAG, "Account Type: " + accountType);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        switch (requestCode) {
            case PERMISSION_REQUEST_CODE:
                Log.d(TAG, "Request Permissions Result:");

                boolean isGranted = grantResults[0] == PackageManager.PERMISSION_GRANTED;

                String result = (isGranted) ? "PERMISSION_GRANTED" : "PERMISSION_DENIED";
                Log.d(TAG, "- " + permissions[0] + ": " + result);

                if (isGranted) {
                    chooseAccountIntent();
                }
                break;
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == 200) {
            if (data != null) {
                Bundle extras = data.getExtras();
                final String accountName = extras.getString(AccountManager.KEY_ACCOUNT_NAME);
                final String accountType = extras.getString(AccountManager.KEY_ACCOUNT_TYPE);
                Log.d(TAG, "Account Name: " + accountName);
                Log.d(TAG, "Account Type: " + accountType);

                readAccount();

                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        // Get Auth Token for Google
                        AccountManager accountManager = AccountManager.get(MainActivity.this);
                        AccountManagerFuture<Bundle> accountManagerFuture = accountManager.getAuthToken(new Account(accountName, accountType), "ah", null, MainActivity.this, null, null);
                        Bundle authTokenBundle = null;
                        try {
                            authTokenBundle = accountManagerFuture.getResult();
                        } catch (OperationCanceledException e) {
                            e.printStackTrace();
                        } catch (IOException e) {
                            e.printStackTrace();
                        } catch (AuthenticatorException e) {
                            e.printStackTrace();
                        }
                        if (authTokenBundle != null) {
                            String authToken = authTokenBundle.get(AccountManager.KEY_AUTHTOKEN).toString();
                            Log.d(TAG, "Account Auth Token: " + authToken);
                        } else {
                            Log.d(TAG, "Account Auth Token is null");
                        }
                    }
                }).start();



            }
        }
    }
}

File: AccountManagerSample/src/main/java/com/daya/android/accounts/sample/MainActivity.kt
package com.daya.android.accounts.sample

import android.accounts.Account
import android.accounts.AccountManager
import android.accounts.AccountManager.KEY_ACCOUNT_NAME
import android.accounts.AccountManager.KEY_ACCOUNT_TYPE
import android.accounts.AuthenticatorException
import android.accounts.OperationCanceledException
import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import com.daya.android.common.DayaLog
import java.io.IOException

class MainActivity : AppCompatActivity() {

    private lateinit var accountManager: AccountManager

    private val accountType
        get() = getString(R.string.account_type)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        accountManager = AccountManager.get(this)

        findViewById<Button>(R.id.login_button).setOnClickListener {
            addAccount()
        }

        findViewById<Button>(R.id.choose_account_button).setOnClickListener {
            chooseAccount()
        }

        findViewById<Button>(R.id.show_accounts_button).setOnClickListener {
            showAccounts()
        }
    }

    private fun addAccount() {
        val addAccountOptions = Bundle().apply {
            putBoolean("SelectAccount", true)
        }
        accountManager.addAccount(accountType, null, null, addAccountOptions, this, {
            DayaLog.d(TAG, "Add account result.")
            try {
                val bundle = it.result
                DayaLog.d(
                    TAG, """Add account result: 
                        Account name: ${bundle.getString(AccountManager.KEY_ACCOUNT_NAME)}
                        Account type: ${bundle.getString(AccountManager.KEY_ACCOUNT_TYPE)}
                    """.trimMargin()
                )
                val account = Account(
                    bundle.getString(AccountManager.KEY_ACCOUNT_NAME),
                    bundle.getString(AccountManager.KEY_ACCOUNT_TYPE)
                )
                getAuthToken(account)

            } catch (e: AuthenticatorException) {
                e.printStackTrace()
            } catch (e: OperationCanceledException) {
                e.printStackTrace()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }, Handler())
    }

    private fun getAuthToken(account: Account) {
        accountManager.getAuthToken(account, "simpleToken", null, this, {
            val result = it.result
            DayaLog.d(
                TAG, """Get auth token result: 
                        Account name: ${result.getString(AccountManager.KEY_ACCOUNT_NAME)}
                        Account type: ${result.getString(AccountManager.KEY_ACCOUNT_TYPE)}
                        Auth token: ${result.getString(AccountManager.KEY_AUTHTOKEN)}
                    """.trimMargin()
            )
        }, Handler())
    }

    private fun chooseAccount() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            DayaLog.d(TAG, "Account type: $accountType")
            val intent = AccountManager.newChooseAccountIntent(
                null,
                null,
                arrayOf(accountType),
                null,
                null,
                null,
                null
            )
            startActivityForResult(intent, REQUEST_CHOOSE_ACCOUNT)
        }
    }

    private fun showAccounts() {
        val accounts = accountManager.getAccountsByType(accountType)
        DayaLog.d(TAG, "Account size: ${accounts.size}")
        accounts.forEach {
            DayaLog.d(
                TAG, """
                Account type: ${it.type}
                Account name: ${it.name}
            """.trimIndent()
            )
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        if (requestCode == REQUEST_PERMISSION) {
            val accountManager = AccountManager.get(this)
            val accounts = accountManager.getAccountsByType(accountType)
            accounts.forEach {
                Log.d(localClassName + "aaaaa1", it.toString())
            }
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_CHOOSE_ACCOUNT) {
            if (resultCode == Activity.RESULT_OK) {
                DayaLog.d(TAG, "Success to choose account.")
                DayaLog.d(
                    TAG, """
                        Success to choose account: 
                            Account type: ${data?.getStringExtra(KEY_ACCOUNT_TYPE)}
                            Account name: ${data?.getStringExtra(KEY_ACCOUNT_NAME)}
                    """.trimIndent()
                )

            } else {
                DayaLog.d(TAG, "Failed to choose account.")
            }

        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    companion object {
        private const val TAG = "MainActivity"
        private const val REQUEST_CHOOSE_ACCOUNT = 100
        private const val REQUEST_PERMISSION = 101
    }
}


